plugin/builtin/sortordertransformer/SortOrderTransformer.go

```diff
@@ -0,0 +1,225 @@
+	// Default to FIFO sort, aka no-op.
+	if p.SortOptions == nil {
+		p.SortOptions = &types.SortOptions{
+			Order: types.FIFOSortOrder,
+		}
+	}
+
+	// If legacy sort is selected and no options are given, default to
+	// hardcoded order.
+	if p.SortOptions.Order == types.LegacySortOrder && p.SortOptions.LegacySortOptions == nil {
+		p.SortOptions.LegacySortOptions = &types.LegacySortOptions{
+			OrderFirst: defaultOrderFirst,
+			OrderLast:  defaultOrderLast,
+		}
+	}
+}
+
+func (p *plugin) validate() error {
+	// Check valid values for SortOrder
+	if p.SortOptions.Order != types.FIFOSortOrder && p.SortOptions.Order != types.LegacySortOrder {
+		return errors.Errorf("the field 'sortOptions.order' must be one of [%s, %s]",
+			types.FIFOSortOrder, types.LegacySortOrder)
+	}
+
+	// Validate that the only options set are the ones corresponding to the
+	// selected sort order.
+	if p.SortOptions.Order == types.FIFOSortOrder &&
+		p.SortOptions.LegacySortOptions != nil {
+		return errors.Errorf("the field 'sortOptions.legacySortOptions' is"+
+			" set but the selected sort order is '%v', not 'legacy'",
+			p.SortOptions.Order)
+	}
+	return nil
+}
+
+func (p *plugin) Transform(m resmap.ResMap) (err error) {
+
+	p.applyDefaults()
+	err = p.validate()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Sort
+	resources := m.Resources()
+	if p.SortOptions.Order == types.LegacySortOrder {
+		s := newLegacyIDSorter(resources, p.SortOptions.LegacySortOptions)
+		sort.Sort(s)
+		// Clear and refill with the correct order
+		m.Clear()
+		for _, r := range resources {
+			m.Append(r)
+		}
+	}
+	return nil
+}
+
+// Code for legacy sorting.
+// Legacy sorting is a "fixed" order sorting maintained for backwards
+// compatibility.
+
+// legacyIDSorter sorts resources based on two priority lists:
+// - orderFirst: Resources that should be placed in the start, in the given order.
+// - orderLast: Resources that should be placed in the end, in the given order.
+type legacyIDSorter struct {
+	resources  []*resource.Resource
+	options    *types.LegacySortOptions
+	typeOrders map[string]int
+}
+
+func newLegacyIDSorter(
+	resources []*resource.Resource,
+	options *types.LegacySortOptions) *legacyIDSorter {
+
+	// Precalculate a resource ranking based on the priority lists.
+	var typeOrders = func() map[string]int {
+		m := map[string]int{}
+		for i, n := range options.OrderFirst {
+			m[n] = -len(options.OrderFirst) + i
+		}
+		for i, n := range options.OrderLast {
+			m[n] = 1 + i
+		}
+		return m
+	}()
+	return &legacyIDSorter{
+		resources:  resources,
+		options:    options,
+		typeOrders: typeOrders,
+	}
+}
+
+var _ sort.Interface = legacyIDSorter{}
+
+func (a legacyIDSorter) Len() int { return len(a.resources) }
+func (a legacyIDSorter) Swap(i, j int) {
+	a.resources[i], a.resources[j] = a.resources[j], a.resources[i]
+}
+func (a legacyIDSorter) Less(i, j int) bool {
+	if !a.resources[i].GetGvk().Equals(a.resources[j].GetGvk()) {
```

Comment:
Even though we're not leveraging an `IdSlice` type to define the sorting anymore (because what we want is bespoke to this file, and we need to handle the additional state of typeOrders), I think we do want to retain the optimization of sorting `resid.ResId` instead of `resource.Resource`, because of this line: `resource.GetGVK()` is somewhat expensive, so we don't want to do it over and over during the comparisons. The conversion to `[]resid.ResId` basically precomputes this for us. I think switching `legacyIDSorter` to this internally should be pretty straighforward, i.e. start with:

```go
type legacyIDSorter struct {
	resIds     []resid.ResId // change to this
	options    *types.LegacySortOptions
	typeOrders map[string]int
}

// and do this in Transform, like the old transformer did
resources := make([]*resource.Resource, m.Size())
		ids := m.AllIds()
		s := newLegacyIDSorter(ids, p.SortOptions.LegacySortOptions)
		sort.Sort(s)
		// Clear and refill with the correct order
		for i, id := range ids {
			resources[i], err = m.GetByCurrentId(id)
			if err != nil {
				return errors.Wrap(err, "expected match for sorting")
			}
		}
		m.Clear()
		for _, r := range resources {
			m.Append(r)
		}
```

Reply:
That makes sense! I'll change the code to do that instead.